Add extension cookbook (#14168)
quarkusio/quarkus: docs/src/main/asciidoc/extension-cookbook.adoc

= Extension Cookbook

== Context & Objective

> The purpose of this article is to be more dev oriented and less api oriented to explain how to produce extension for quarkus.

Reviewer: ```suggestion
The purpose of this cookbook is to be more dev oriented and less API oriented to explain how to produce extensions for Quarkus.
```

So it will not detail how to generate BuildItem or use substitution, but it will explain the workflow to handle extension.

> First an extension is a module which will be run by quarkus during build time to generate code and inject it.

Reviewer: Make sure to use `Quarkus` instead of `quarkus`

> It is a kind of pre-compilation.

Reviewer: This sentence is a bit loose, maybe move to the previous sentence?

== Organisation

There are 2 projects:

> - runtime will contain all class loaded on runtime
> - deployment which contain processor class which will be run during build (augmenter).

Reviewer: ```suggestion
- **runtime**:: Contains all classes loaded during runtime
- **deployment**:: Contains all processor classes which will be run during the build.
```

The processor is the heart of the extension all build step methods are a hierarchy of dependant.
Quarkus automatically detect build items input and output and run then in the right order.

> It mean that sometimes you add an uneeded input just ot be sure it is launched after another method.

Reviewer: ```suggestion
It mean that sometimes you add an unneeded input just to be sure it is launched after another method.
```

> The processor can generated bytecode, read config, and pass some data to runtime thanks to recorder (only simple structure).

Reviewer: ```suggestion
The processor can generate bytecode, read config, and pass some data to runtime thanks to recorder (only simple structure).
```

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
////
= Extension Cookbook

:numbered:
:sectnums:
:sectnumlevels: 4
:toc:

include::./attributes.adoc[]

== Context & Objective

The purpose of this cookbook is to explain for extension developer, a step by step process to build an extension from scratch.

> Other guide are API oriented and this one will not detailed API. For this purpose use other guides such as :

Reviewer: This sentence is confusing

Author: is it better with
It do not explain the quarkus API in detail. For this purpose use other guides such as :
